ethcore/light/src/on_demand/mod.rs

```diff
@@ -403,10 +435,14 @@ impl OnDemand {
 			required_capabilities: capabilities,
 			responses,
 			sender,
-			base_query_index: 0,
-			remaining_query_count: 0,
-			query_id_history: BTreeSet::new(),
-			inactive_time_limit: None,
+			request_guard: RequestGuard::new(
```

Comment:
I really have question about circuit breakers, for instance it seems to require managing a kind of global state (otherwise why would there be a need for mutex), but here aren't we spawning a new circuit breaker state for each request?

I was expecting that the request guard would be a clone of an existing state (like a state stored for on_demand per user or something more global).
I do not really see how we would get success rate aggregate this way. From what I remember, I did store error query history and remaining query count per request but because I was considering errors (there can be multiple error for a request due to empty reply (maybe one of the latest pr does fix that by slashing those kind of peers) but I think the request lifetime ends at the first request success.

I am not sure at which level the circuit breaker is meant to be applied : 
- for a peer and any of its request?
- for on_demand_service globally (probably not)?
- for a peer and a particular kind of query? (the fact that we can have multiple query per request does not make it easy)
- for a kind of request globaly?
- for every single request and its subsequent replies (that is what I feel is currently done, but I may be wrong)

We should probably document that.

It could make sense to use the other 'failure_policy' with backoff (counting error) at a per request level to basically do the same as before (but if we do it we would need to be able to get rid of the circuit safe mutex that becomes useless).

Reply:
The circuit breakers are on per request basis, and yes it will introduce two mutexes per request and I wanted to separate responses and attempts per request to have a simple logic.

I think `failsafe crate` want to be thread-safe and ensure that only one event can modify the state at the time.

Reply:
The thing I do not understand with a per request basis is that the only relevant event that we will be registering into the circuit breaker will be request failure, so won't the failure policy "SuccessRateOverTimeWindows" act like a simple timeout (average success being always 0)? That is why the use of the other policy "ConsecutiveFailure" seems more intuitive to me.

On another hand, if we choose to use back our circuit state like single circuit state for each peers, we would register some successes and I would understand (but I do not expect it to be easy, especially in this future based code). Plus we need to be careful to not mess to much with the random selection of peers to avoid overloading some nodes, but that is another story.

For instance adding this test in response_guard illustrate that (the test is passing) : 

```rust
	#[test]
	fn act_like_timeout() {
		let mut guard = ResponseGuard::new(0.8, Duration::from_secs(5));
		thread::sleep(Duration::from_secs(5));
		assert!(guard.register_error(&ResponseError::Validity(ValidityError::WrongKind)).is_err());

    let mut guard = ResponseGuard::new(0.2, Duration::from_secs(5));
		assert!(guard.register_error(&ResponseError::Unexpected).is_ok());
		thread::sleep(Duration::from_secs(5));
		assert!(guard.register_error(&ResponseError::Validity(ValidityError::WrongKind)).is_err());

    let mut guard = ResponseGuard::new(0.8, Duration::from_secs(5));
    
    for i in 0..1000 {
		  assert!(guard.register_error(&ResponseError::Unexpected).is_ok());
    }
		thread::sleep(Duration::from_secs(5));
		assert!(guard.register_error(&ResponseError::Validity(ValidityError::WrongKind)).is_err());
	}

```

Reply:
> The thing I do not understand with a per request basis is that the only relevant event that we will be registering into the circuit breaker will be request failure, so won't the failure policy "SuccessRateOverTimeWindows" act like a simple timeout (average success being always 0)? That is why the use of the other policy "ConsecutiveFailure" seems more intuitive to me.

Yes, that is correct but I argue that is more intuitive for a user to specify a certain time window then the number of `bad-responses` because the number of peers is dynamic. However, that the first assert is passing was not expected from my side. I thought that the time window would be `reset` here and only be registered after the time window had passed. I need to re-think this maybe total number of failures or consecutive failures are better anyhow.

> single circuit state for each peer

No, that is too fine-grained IMHO